crates/uv-resolver/src/lock.rs

```diff
@@ -2119,6 +2209,64 @@ impl TryFrom<SourceWire> for Source {
     }
 }
 
+/// The source for a registry, which could be a URL or a relative path.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+enum RegistrySource {
+    /// Ex) `https://pypi.org/simple`
+    Url(UrlString),
+    /// Ex) `../path/to/local/index`
+    Path(PathBuf),
+}
+
+impl std::fmt::Display for RegistrySource {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            RegistrySource::Url(url) => write!(f, "{url}"),
+            RegistrySource::Path(path) => write!(f, "{}", path.display()),
+        }
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for RegistrySource {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for Visitor {
+            type Value = RegistrySource;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a valid URL or a file path")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                if split_scheme(value).is_some() {
+                    Ok(
+                        serde::Deserialize::deserialize(serde::de::value::StrDeserializer::new(
+                            value,
+                        ))
+                        .map(RegistrySource::Url)?,
+                    )
+                } else {
+                    Ok(
+                        serde::Deserialize::deserialize(serde::de::value::StrDeserializer::new(
+                            value,
+                        ))
+                        .map(RegistrySource::Path)?,
+                    )
+                }
+            }
+        }
```

Comment:
Not _really_ that happy with this, but I didn't want to introduce separate enum variants for remote vs. local registries.